// neon-init/dist/lib/install.js
import { createApiKeyFromNeonctl, ensureNeonctlAuth } from "./auth.js";
import { getMCPConfig, writeMCPConfig } from "./mcp-config.js";
import { confirm, isCancel, log, spinner } from "@clack/prompts";

//#region src/lib/install.ts
/**
* Installs Neon's MCP Server for a specific editor
* - Cursor: Global config
* - VS Code: Global config (preferred) or workspace config (fallback)
* - Claude CLI: Global config
*/
async function installMCPServerForEditor(homeDir, workspaceDir, editor, apiKey) {
	const { config, configPath } = getMCPConfig(homeDir, workspaceDir, editor);
	const serverKey = editor === "VS Code" ? config.servers : config.mcpServers;
	if (Boolean(serverKey?.Neon)) {
		const response = await confirm({
			message: `Neon MCP Server is already configured for ${editor}. Would you like to reconfigure it? (Y/n)`,
			initialValue: true
		});
		if (isCancel(response)) return "failed";
		if (!response) {
			log.info(`Keeping existing configuration for ${editor}.`);
			return "success";
		}
	}
	const neonServerConfig = {
		type: "http",
		url: "https://mcp.neon.tech/mcp",
		headers: { Authorization: `Bearer ${apiKey}` }
	};
	if (editor === "VS Code") {
		if (!config.servers) config.servers = {};
		config.servers.Neon = neonServerConfig;
	} else {
		if (!config.mcpServers) config.mcpServers = {};
		config.mcpServers.Neon = neonServerConfig;
	}
	try {
		writeMCPConfig(configPath, config);
		return "success";
	} catch (error) {
		log.error(`Failed to write configuration for ${editor}: ${error instanceof Error ? error.message : "Unknown error"}`);
		return "failed";
	}
}
/**
* Installs Neon's MCP Server
* - Cursor: Global config
* - VS Code: Global config (preferred) or workspace config (fallback)
* - Claude CLI: Global config
*/
async function installMCPServer(homeDir, workspaceDir, selectedEditors) {
	const results = /* @__PURE__ */ new Map();
	const authSpinner = spinner();
	authSpinner.start("Authenticating...");
	if (!await ensureNeonctlAuth()) {
		authSpinner.stop("Authentication failed");
		for (const editor of selectedEditors) results.set(editor, "failed");
		return results;
	}
	authSpinner.stop("Authentication successful ✓");
	const apiKey = await createApiKeyFromNeonctl();
	if (!apiKey) {
		log.error("Could not create API key after authentication.");
		log.info("You can manually create one at: https://console.neon.tech/app/settings/api-keys");
		for (const editor of selectedEditors) results.set(editor, "failed");
		return results;
	}
	for (const editor of selectedEditors) {
		const status = await installMCPServerForEditor(homeDir, workspaceDir, editor, apiKey);
		results.set(editor, status);
	}
	return results;
}

//#endregion
export { installMCPServer };
